### PR TITLE
feat: new HTMLDirective design

### PR DESCRIPTION
**Pull request by EisenbergEffect**
# Pull Request

## 📖 Description

This PR improve the API of `HTMLDirective`. The changes are as follows:

* Decouples directives and behaviors.
* Directives can now add 0-N behaviors when they generate HTML.
* `createPlaceholder` is changed to `createHTML` and now receives a method for adding behavior factories.
* Behavior application is based on unique ids that are generated when a directive adds the behavior factory.
* Base classes removed for directive. Everything is based on interfaces and registration now.
* Directives are now registered with `HTMLDirective.define` API
* You can also decorate a class to register it as a directive with `@htmlDirective`
* Directives can now indicate through their registration options that they are "aspected" to get access to the attribute/property/event name and the aspect type before they generate HTML

### Examples

Here's an example of the most basic `HTMLDirective`:

```ts
@htmlDirective()
class HelloWorldDirective implements HTMLDirective {
  createHTML() {
    return "Hello World";
  }
}

// alternative to the decorator
// HTMLDirective.define(HelloWorldDirective); 
```

If a directive wants to add behavior factories, then you do that like this:

```ts
@htmlDirective({ aspected: true })
class DirectiveWithBehavior implements HTMLDirective {
  sourceAspect: string; // e.g. :someProperty
  targetAspect: string; // e.g. someProperty
  aspectType: number; // e.g. Aspect.property

  createHTML(add: AddViewBehaviorFactory) {
    // Create an interpolation placeholder associated with the custom behavior
    // This directive is aspected so it will also receive info indicating what it's
    // being applied to.
    return Markup.interpolation(add(new SomeBehaviorFactoryIWantToAdd(this)));
  }
}

// alternative to the decorator
// HTMLDirective.define(DirectiveWithBehavior, { aspected: true })); 
```

You can now even add multiple behaviors:

```ts
@htmlDirective()
class DirectiveWithMultipleBehaviors implements HTMLDirective {
  createHTML(add: AddViewBehaviorFactory) {
    const placeholder1 = Markup.interpolation(add(new SomeBehaviorFactoryIWantToAdd(this)));
    const placeholder2 = Markup.interpolation(add(new AnotherBehaviorFactoryIWantToAdd(this)));
    // each behavior will control the individual text node represented by the placeholder
    return `Hello ${placeholder1} World ${placeholder2}`;
  }
}
```

Of course, for optimization purposes, the directive can also serve as the factory by adding itself and implementing the `ViewBehaviorFactory` interface. Even the factory can serve as the behavior by implementing the `ViewBehavior` interface. Most out-of-the-box directives use this pattern in order to avoid unnecessarily creating extra objects. However, now these concepts are properly decoupled so that more advanced scenarios can be accomplished.

### 🎫 Issues

This is the first half of #5799 

## 👩‍💻 Reviewer Notes

A good bit of side-effects on this one so another pair of eyes throughout is welcome.

## 📑 Test Plan

All existing tests are updated and passing. I added a couple more that test new interpolation scenarios and the registration API.

### General

- [x] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

## ⏭ Next Steps

After this PR is merged I'll be making a similar set of changes to the `CSSDirective`.

### CODE REVIEW

**Comment by nicholasrice** [comment]
@EisenbergEffect I think you need to rebase this branch because I rebased fast-element-2 off master last week. It might be easiest to just cherry-pick the commits you want to a new branch, cause the rebase wasn't exactly straight-forward.

**Comment by chrisdholt** [comment]
@eisenbergeffect looks like this needs to be rebased - I'd jump in now but I think it'll be easier once that's done :)

**Comment by EisenbergEffect** [comment]
Rebased.

@nicholasrice The tests for the new reflect directive are broken for some reason. I'm going to try to figure that out but may need your help.

**Comment by EisenbergEffect** [comment]
Ok, this is ready for review.

File: packages/web-components/fast-ssr/src/template-parser/template-parser.spec.ts [hunk 1]
```diff
@@ -28,22 +36,22 @@ test.describe("parseTemplateToOpCodes", () => {
...
             const input = html`Hello ${() => "World"}.`;
 
             const codes = parseTemplateToOpCodes(input);
-            const code = codes[1] as DirectiveOp;
+            const code = codes[1] as ViewBehaviorFactoryOp;
             expect(codes.length).toBe(3);
-            expect(code.type).toBe(OpType.directive);
+            expect(code.type).toBe(OpType.viewBehaviorFactory);
     });
     test("should sandwich directive ops between text ops when binding native element content", () => {
```
Comment (nicholasrice) [inline packages/web-components/fast-ssr/src/template-parser/template-parser.spec.ts:49]:
```suggestion
    test("should sandwich viewBehaviorFactory ops between text ops when binding native element content", () => {
```

File: packages/web-components/fast-ssr/src/template-parser/template-parser.spec.ts [hunk 2]
```diff
@@ -28,22 +36,22 @@ test.describe("parseTemplateToOpCodes", () => {
     })
     test("should emit a directive op from a binding", () => {
```
Comment (nicholasrice) [inline packages/web-components/fast-ssr/src/template-parser/template-parser.spec.ts:37]:
```suggestion
    test("should emit a viewBehaviorFactory op from a binding", () => {
```

File: packages/web-components/fast-ssr/test/example.spec.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { test, expect, ElementHandle } from '@playwright/test';
+import fastSSR from "../src";
+
+test("example module test", async () => {
+	expect(fastSSR).toBe("fast-ssr");
+})
```
Comment (nicholasrice) [inline packages/web-components/fast-ssr/test/example.spec.ts:6]: I believe this whole test directory was removed, is this an artifact of the rebase?

Comment (EisenbergEffect) [reply packages/web-components/fast-ssr/test/example.spec.ts:6]: Oops. Probably. It was a very long rebase. Is it something you can add a commit to bring it back easily?

File: packages/web-components/fast-foundation/src/directives/reflect-attributes.ts [hunk 4]
```diff
@@ -79,14 +88,17 @@ class ReflectAttrBehavior extends StatelessAttachedAttributeDirective<string[]>
...
-        this.target = targets[this.targetId] as HTMLElement;
-        AttributeReflectionSubscriptionSet.getOrCreateFor(source).subscribe(this);
+        const subscription = AttributeReflectionSubscriptionSet.getOrCreateFor(source);
+        subscription[this.id] = targets[this.nodeId];
+        subscription.subscribe(this);
 
         // Reflect any existing attributes because MutationObserver will only
         // handle *changes* to attributes.
         if (source.hasAttributes()) {
+            console.log("has attributes");
```
Comment (nicholasrice) [inline packages/web-components/fast-foundation/src/directives/reflect-attributes.ts:98]:
```suggestion
```

File: packages/web-components/fast-foundation/src/directives/reflect-attributes.ts [hunk 5]
```diff
@@ -79,14 +88,17 @@ class ReflectAttrBehavior extends StatelessAttachedAttributeDirective<string[]>
...
             for (let i = 0; i < source.attributes.length; i++) {
-                this.handleChange(source, source.attributes[i].name);
+                console.log("sending attribute");
```
Comment (nicholasrice) [inline packages/web-components/fast-foundation/src/directives/reflect-attributes.ts:100]:
```suggestion
```

**Review by nicholasrice — APPROVED** [review]

File: packages/web-components/fast-ssr/test/example.spec.ts (hunk 3 above)
Comment (EisenbergEffect) [reply packages/web-components/fast-ssr/test/example.spec.ts]: Removed this folder.

**Pull request merged**